Mage.Sets/src/mage/cards/g/GoldberryRiverDaughter.java

```diff
@@ -0,0 +1,183 @@
+    GoldberryRiverDaughterToEffect() {
+        super(Outcome.Neutral);
+        staticText = "Move one or more counters from Goldberry onto another target permanent you control. If you do, draw a card.";
+    }
+
+    private GoldberryRiverDaughterToEffect(final GoldberryRiverDaughterToEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public GoldberryRiverDaughterToEffect copy() {
+        return new GoldberryRiverDaughterToEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        Permanent toPermanent = game.getPermanent(source.getFirstTarget());
+        Permanent fromPermanent = game.getPermanent(source.getSourceId());
+
+        if (fromPermanent == null
+                || toPermanent == null
+                || controller == null
+                || fromPermanent.getCounters(game).size() == 0) {
+            return false;
+        }
+
+        List<Counter> counters = new ArrayList<>(fromPermanent.getCounters(game).values());
+        counters.sort((c1, c2) -> c1.getName().compareTo(c2.getName()));
+
+        List<MultiAmountMessage> messages = counters.stream()
+                .map(c -> new MultiAmountMessage(c.getName() + " (" + c.getCount() + ")", 0, c.getCount()))
+                .collect(Collectors.toList());
+        int max = messages.stream().map(m -> m.max).reduce(0, Integer::sum);
+
+        int total;
+        List<Integer> choices;
+        do {
+            choices = controller.getMultiAmountWithIndividualConstraints(Outcome.Neutral, messages, 1,
+                    max, MultiAmountType.COUNTERS, game);
+
+            total = choices.stream().reduce(0, Integer::sum);
+        } while (total < 1);
+
+        // Move the counters. Make sure some counters were actually moved.
+        boolean movedCounters = false;
+        for (int i = 0; i < choices.size(); i++) {
+            Integer amount = choices.get(i);
+
+            if (amount > 0) {
+                String counterName = counters.get(i).getName();
+
+                movedCounters |= toPermanent.addCounters(
+                        CounterType.findByName(counterName).createInstance(amount),
+                        source,
+                        game);
+                fromPermanent.removeCounters(counterName, amount, source, game);
+                game.informPlayers(
+                        controller.getLogName() + "moved " +
+                                amount + " " +
+                                counterName + " counter" + (amount > 1 ? "s" : "") +
+                                " from " + fromPermanent.getLogName() +
+                                "to " + toPermanent.getLogName() + ".");
+            }
+        }
+
+        // If some counters were actually moved, draw a card
+        if (movedCounters) {
+            controller.drawCards(1, source, game);
+        }
+        return false;
```

Comment:
should return true? though probably doesn't matter

Reply:
I'm going to be honest, I don't know what the return value for effects is supposed to signify. It's always seemed nonsensical to me.

Reply:
I'm pretty sure the return value signifies whether the effect actually happened.

Reply:
It looks like the return value is only used for effects which are part of mana abilities and special actions. In which case it isn't really clear to me when effects like this one should or should not return false (does it even matter?).

Reply:
Since this is an inner custom effect that nothing else will call, it does not matter. But the convention is as Grath describes, to return `true` if something happens and `false` if nothing happens. It can matter for common effects that have their `apply()` method called within the `apply()` method of another effect (either a subclass calling super, or just creating a new effect to use its logic) and trying to take action based on whether or not that effect succeeded. Of course, whenever attempting this it is important to double check that the logic will work as intended. Because it so often doesn't matter, there are lots of effects that don't adhere properly to the convention with no ill effects other than confusing developers.